pkg/cli/parse_quotas.go

```diff
@@ -21,102 +21,237 @@ package cli
+type resourceUnits map[string]map[string]limes.Unit
+
+// baseUnitsSetter is the interface type that is implemented by different
+// hierarchies.
+type baseUnitsSetter interface {
+	setBaseUnits(*resourceUnits)
+}
+
+// ParseRawQuotas parses the raw quota values given at the command line to a
+// Quotas map.
+func ParseRawQuotas(s baseUnitsSetter, rq *RawQuotas) (*Quotas, error) {
+	q := &Quotas{}
+
+	type userInput struct {
+		service  string
+		resource string
+		value    string
+		unit     limes.Unit
+		comment  string
 	}
-	srv := srvRes[0]
-	tmp.Name = srvRes[1]
-
-	// separate quota value and comment (if one was given)
-	valCom := strings.SplitN(idfVal[1], ":", 2)
-	if len(valCom) > 1 {
-		if valCom[1] != "" {
-			tmp.Comment = strings.TrimSpace(valCom[1])
+
+	userInputs := make([]userInput, 0, len(*rq))
+	resUnits := make(resourceUnits)
+
+	// validate raw quota values
+	for _, rqInList := range *rq {
+		var input userInput
+
+		// separate the quota value from its identifier
+		idfVal := strings.SplitN(rqInList, "=", 2)
+		if len(idfVal) != 2 {
+			return nil, fmt.Errorf("expected a quota in the format: service/resource=value(unit), got '%s'", rqInList)
 		}
-	}
 
-	// separate quota's value from its unit (if one was given)
-	rx := regexp.MustCompile(`^([0-9]+)([A-Za-z]+)?$`)
-	match := rx.MatchString(valCom[0])
-	if !match {
-		return fmt.Errorf("expected a quota value with optional unit in the format: 123Unit, got '%s'", valCom[0])
-	}
+		// separate service and resource
+		srvRes := strings.SplitN(idfVal[0], "/", 2)
+		if len(srvRes) != 2 {
+			return nil, fmt.Errorf("expected service/resource, got '%s'", idfVal[0])
+		}
+		input.service = srvRes[0]
+		input.resource = srvRes[1]
 
-	// rxMatchedList: []string{"entire regex matched string", "quota value", "unit (empty, if no unit given)"}
-	rxMatchedList := rx.FindStringSubmatch(valCom[0])
-	intVal, err := strconv.ParseInt(rxMatchedList[1], 10, 64)
-	if err != nil {
-		return fmt.Errorf("could not parse quota value: '%s'", rxMatchedList[1])
-	}
-	tmp.Value = intVal
-
-	tmp.Unit = limes.UnitNone
-	if rxMatchedList[2] != "" {
-		switch rxMatchedList[2] {
-		case "B":
-			tmp.Unit = limes.UnitBytes
-		case "KiB":
-			tmp.Unit = limes.UnitKibibytes
-		case "MiB":
-			tmp.Unit = limes.UnitMebibytes
-		case "GiB":
-			tmp.Unit = limes.UnitGibibytes
-		case "TiB":
-			tmp.Unit = limes.UnitTebibytes
-		case "PiB":
-			tmp.Unit = limes.UnitPebibytes
-		case "EiB":
-			tmp.Unit = limes.UnitExbibytes
-		default:
-			return fmt.Errorf("acceptable units: ['B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB'], got '%s'", rxMatchedList[2])
+		// separate quota value and comment (if one was given)
+		valCom := strings.SplitN(idfVal[1], ":", 2)
+		if len(valCom) > 1 {
+			if valCom[1] != "" {
+				input.comment = strings.TrimSpace(valCom[1])
+			}
+		}
+
+		// separate quota's value from its unit (if one was given)
+		rx := regexp.MustCompile(`^([0-9]*.[0-9]+)([A-Za-z]+)?$`)
```

Comment:
1. `MustCompile` in a loop is not nice. The regex should be compiled once (e.g. in a global variable assignment) and then reused.

2. The `.` in the regex must be escaped, otherwise it matches all characters, so e.g. `21g65MiB` would be matched. Also, the dot should be matched with a `?` quantor, otherwise numbers without dots such as `10MiB` would not be matched. Please include unit tests for these cases.